F/timestamp based elections

p4u opened the pull request.

p4u committed:
vochain: introduce timestamp based elections

Elections can be created using UNIX timestamps
instead of block number. The timestamp is taken
from the consensus block header.

Signed-off-by: p4u <pau@dabax.net>

p4u committed:
vochain: adapt components and tests to timestamp based blockchain

Signed-off-by: p4u <pau@dabax.net>

p4u committed:
Merge remote-tracking branch 'origin/main' into f/timestamp_based_elections

Signed-off-by: p4u <pau@dabax.net>

p4u committed:
remove unused validators from stage

Signed-off-by: p4u <pau@dabax.net>

p4u committed:
indexer: adapt to timestamp based blockchain

Signed-off-by: p4u <pau@dabax.net>

p4u committed:
vochain: log processId on deprecated elections based on block count

Signed-off-by: p4u <pau@dabax.net>

p4u committed:
Merge branch 'main' into f/timestamp_based_elections

jordipainan commented on apiclient/election.go:
```diff
@@ -95,28 +95,18 @@ func (c *HTTPclient) NewElection(description *api.ElectionDescription) (types.He
+	var startTime, duration uint32
+	if description.StartDate.IsZero() {
+		// if start date is empty, start the election immediately.
+		startTime = 0
```
`startTime` is already `0` at this point since it is initialized at line 102 and not modified. You can delete `startTime = 0`

jordipainan commented on cmd/end2endtest/helpers.go:
```diff
@@ -770,3 +771,24 @@ func (t *e2eElection) registerAnonAccts(voterAccounts []*ethereum.SignKeys) erro
+
+// logElection prints the election description in the log.
+func logElection(e *vapi.Election) {
+	log.Debugw("election description",
```
IMO the `election description` message is misleading as one can expect some actual description of the election.
I suggest to change it for `election data`. Change the comment accordingly too.

jordipainan commented on cmd/node/main.go:
```diff
@@ -165,7 +165,7 @@ func loadConfig() *config.Config {
 	flag.StringSlice("ipfsConnectPeers", []string{},
 		"use custom ipfsconnect peers/bootnodes for accessing the DHT (comma-separated)")
-	flag.String("archiveURL", types.ArchiveURL, "enable archive retrival from the given IPNS url")
+	flag.String("archiveURL", types.ArchiveURL, "enable archive retrival from the given IPNS url (set \"none\" to disable)")
```
Why not `leave it unset to disable` ?

p4u replied:
Because if uset by flag but present on the YML file or somewhere else, it does not work.

jordipainan commented on dockerfiles/testsuite/env.gateway0:
```diff
@@ -12,3 +12,4 @@ VOCDONI_METRICS_ENABLED=True
 VOCDONI_METRICS_REFRESHINTERVAL=5
 VOCDONI_CHAIN=dev
 VOCDONI_SIGNINGKEY=e0f1412b86d6ca9f2b318f1d243ef50be23d315a2e6c1c3035bc72d44c8b2f90 # 0x88a499cEf9D1330111b41360173967c9C1bf703f
+VOCDONI_ARCHIVEURL=none
```
Cannot use an empty string ?

p4u replied:
nop

jordipainan commented on util/helpers.go:
```diff
@@ -67,3 +68,8 @@ func BuildURL(parts ...string) (string, error) {
 }
+
+// TimestampToTime converts a uint32 timestamp to a time.Time object.
+func TimestampToTime(timestamp uint32) time.Time {
```
Since `timestamp` will be passed at execution time, we need to check  for a possible overflow here. If a big enough value is passed it will be truncated and will produce an unexpected time.

p4u replied:
time.Unix() is int64 on Golang, what do you mean by check possible overflow?

jordipainan commented on vochain/indexer/db/models.go:
```diff
@@ -25,11 +26,10 @@ type Block struct {
 	EndDate            time.Time
-	BlockCount         int64
+	ManuallyEnded      bool
+	BlockCount         sql.NullInt64
```
Do we really need to use this type ?

p4u replied:
You are probably seeing the diff changes from the feature branch to the main. This change is not from this PR.

jordipainan replied:
Seems so, ignore then my last comment, regarding mixing changes. Sorry about that.

jordipainan commented on vochain/indexer/process.go:
```diff
@@ -144,21 +144,14 @@ func (idx *Indexer) newEmptyProcess(pid []byte) error {
-		StartBlock:        int64(p.StartBlock),
-		EndBlock:          int64(p.BlockCount + p.StartBlock),
-		BlockCount:        int64(p.BlockCount),
+		StartDate:         time.Unix(int64(p.StartTime), 0),
```
Why not using `TimestampToTime` function recently created ?
Same for all similar operations.

jordipainan commented on vochain/state/electionprice/electionprice.go:
```diff
@@ -164,6 +167,12 @@ func (p *Calculator) Price(params *ElectionParameters) uint64 {
 
+	// If the election duration is specified in seconds, convert it to blocks
+	// This is a temporary solution until the block duration support is removed.
+	if params.ElectionDurationSeconds > 0 {
```
Is this still required?, asking because new created elections will be working with time and not blocks.

p4u replied:
We still calculate the price in blocks because the formula is based on blocks. Change to seconds would require changing the full formula, and I let this for a future PR (since it requires also simulations and so on).

jordipainan commented on vochain/state/state.go:
```diff
@@ -420,6 +430,32 @@ func (v *State) SetHeight(height uint32) {
 }
 
+// SetTimestamp sets the timestamp for the current (not committed) block.
+func (v *State) SetTimestamp(timestamp uint32) error {
```
We should probably check for uint32 overflows here too

p4u replied:
By overflow you mean calling SetTimestamp with a uint32 value higher than  2^32 -1 ? If that is what you mean, there is no way to check the overflow here, the uint32 will already have an "overflow" value at this point. I.e imagine I call `SetTimestamp(-1)` (underflow), the function will receive timestamp=2^32 -1 which is a valid uint32

jordipainan replied:
You're right. I was thinking about handling possible unwanted values, more of a design consideration.
The responsibility basically lies in ensuring that the function is used correctly.
My thoughts were on modifying the function to something like:

```golang
func TimestampToTime(timestamp uint32) (time.Time, error) {
    if timestamp < 0 || timestamp > math.MaxUint32 {
        // return error
    }
    return time.Unix(timestamp, 0), nil
}
```

But it does not make much sense, Thank  you for the clarifications.

jordipainan reviewed (commented):
Thank you for the PR! 

Changing from block calculation to timestamps is going make the vochain better and we will get rid of annoying block to time calculations that have been a pain in the past ^^ 

Left some comments :)

I also want to mention that seems that this PR does not only changes the way elections work (from blocks to timestamps), but it also includes changes on IST and other small things. Please include changes like the ones commented in other PRs in the future, that way the reviewers can focus on the scope of the PR and can be more effective.

p4u committed:
Merge remote-tracking branch 'origin/main' into f/timestamp_based_elections

Signed-off-by: p4u <pau@dabax.net>

p4u committed:
end2endtest: add test parallelization and number of runs

p4u committed:
e2etest: add new test for checking election time bounds

Signed-off-by: p4u <pau@dabax.net>

p4u committed:
vochain: do not return error on CheckTx, just Code=1

mvdan committed:
vochain/indexer: add a test that restores an older backup

This allows us to test our SQL migrations with real data from staging.
In particular, we need to now write a migration that moves data
from two columns, to be deleted, to one new column.
That is non-trivial enough to need a test.

Running the test on the PR changing the existing migrations
in a backwards-incompatible way breaks the test, as expected:

    --- FAIL: TestRestoreBackupAndMigrate (0.04s)
        migrations_test.go:46:
            error:
              got non-nil error
            got:
              e"error preparing query CreateProcess: table processes has no column named manually_ended"
            stack:
              /home/mvdan/src/voc/vochain/indexer/migrations_test.go:46
                qt.Assert(t, err, qt.IsNil)

The sqlite backup file is about 17MiB, but just under 5MiB with
zstd -15, which seems reasonable enough for a git testdata file.
We could trim it down, but we also want all that real data to test with.

mvdan committed:
vochain/indexer: reset the migrations changes

To un-break the existing databases and fix the new test.
The next commit will re-do this as a new migration.

mvdan committed:
vochain/indexer: add a migration for processes.manually_ended

Note that, unlike Pau's original change to older migrations,
the block_count column is dropped as well since it's no longer used.

mvdan committed:
vochain/indexer: re-run go generate

It seems like this wasn't done with some recent sql query changes.
